corehq/util/tests/test_couch.py

```diff
@@ -76,3 +77,121 @@ def test_get_document_or_404_success(self):
         """
         doc = get_document_or_404(MockModel, 'ham', '123')
         self.assertEqual(doc, {'wrapped': {'_id': '123', 'domain': 'ham', 'doc_type': 'MockModel'}})
+
+
+class LoggingDB(object):
+    def __init__(self):
+        self.docs_saved = []
+        self.num_writes = 0
+
+    def bulk_save(self, docs):
+        self.docs_saved.extend(docs)
+        self.num_writes += 1
+        return [{'id': 'unique_id'} for doc in docs]
+
+
+class IterDBSimpleTest(SimpleTestCase):
+    def test_number_of_calls(self):
+        db = LoggingDB()
+        with IterDB(db, chunksize=50) as iter_db:
+            all_docs = range(105)
+            for doc in all_docs:
+                iter_db.save(doc)
+
+        # the saver should have cleared out all the docs
+        self.assertEqual(iter_db.to_save, [])
+        # The db should have saved all the docs
+        self.assertEqual(db.docs_saved, all_docs)
+        # 105 docs in chunks of 50 should be 3 writes
+        self.assertEqual(db.num_writes, 3)
+
+
+class IterDBTest(TestCase):
+    def setUp(self):
+        self.domain = "TEST"
+        # use Group just 'cause it's a simple model,
+        # and I can't create new models within tests
+        self.db = Group.get_db()
+        self.groups = []
+        for i in range(11):
+            group = Group(domain=self.domain, name="test{}".format(i), index=i)
+            group.save()
+            self.groups.append(group)
+
+    def tearDown(self):
+        for group in Group.by_domain(self.domain):
+            group.delete()
+
+    def test_normal_usage(self):
+        with IterDB(self.db, chunksize=5) as iter_db:
+            for group in self.groups:
+                group['test_property'] = True
+                iter_db.save(group)
+        for group in self.groups:
+            new = self.db.get(group._id)
+            self.assertTrue(new.get('test_property', False))
+
+    def test_conflicted_doc(self):
+        old = self.groups[3]
+        new = Group.get(old._id)
+        new.name = "bwahahaha"
+        new.save()
+        self.assertEqual(old._id, new._id)
+        self.assertNotEqual(old._rev, new._rev)
+        # now if you try to save 'old', it should conflict
+        with IterDB(self.db, chunksize=5) as iter_db:
+            for group in self.groups:
+                iter_db.save(group)
+
+        should_succeed = [g._id for g in self.groups if g._id != old._id]
+        self.assertEqual(should_succeed, iter_db.saved_ids)
+
+        self.assertEqual([old._id], iter_db.error_ids)
```

Comment:
Sorry, not able to understand how `iter_update` gets out of recursion, since it's going to call itself if there are `error_ids`?